File: go/tool_event_golden_vectors_test.go
// Cross-language ToolEvent envelope round-trip fixture test (plan T-2).
//
// Loads `net/crates/net/tests/cross_lang_tool_formats/tool_event_vectors.json`
// and asserts that for each case the Go ToolEvent round-trips
// through encoding/json byte-equal to the wire shape. Matches the
// Rust / Node / Python verifiers.

package net

import (
	"encoding/json"
	"os"
	"path/filepath"
	"reflect"
	"runtime"
	"testing"
)

type fixtureEventCase struct {
	Name       string          `json:"name"`
	Wire       json.RawMessage `json:"wire"`
	IsTerminal bool            `json:"is_terminal"`
}

type fixtureEvents struct {
	Cases []fixtureEventCase `json:"cases"`
}

func eventFixturePath(t *testing.T) string {
	t.Helper()
	_, thisFile, _, ok := runtime.Caller(0)
	if !ok {
		t.Fatal("runtime.Caller failed")
	}
	dir := filepath.Dir(thisFile)
	// go/ lives at the repo root; fixtures live at
	// net/crates/net/tests/cross_lang_tool_formats/.
	return filepath.Join(dir, "..", "net", "crates", "net", "tests", "cross_lang_tool_formats", "tool_event_vectors.json")
}

func loadEventFixture(t *testing.T) fixtureEvents {
	t.Helper()
	raw, err := os.ReadFile(eventFixturePath(t))
	if err != nil {
		t.Fatalf("read fixture: %v", err)
	}
	var f fixtureEvents
	if err := json.Unmarshal(raw, &f); err != nil {
		t.Fatalf("unmarshal fixture: %v", err)
	}
	return f
}

func TestToolEventRoundTripMatchesGoldenVectors(t *testing.T) {
	f := loadEventFixture(t)
	if len(f.Cases) == 0 {
		t.Fatal("no cases in fixture")
	}
	for _, c := range f.Cases {
		c := c
		t.Run(c.Name, func(t *testing.T) {
			// Deserialize wire → ToolEvent.
			var event ToolEvent
			if err := json.Unmarshal(c.Wire, &event); err != nil {
				t.Fatalf("unmarshal wire: %v", err)
			}

			// IsTerminal contract.
			if event.IsTerminal() != c.IsTerminal {
				t.Errorf("IsTerminal() = %v, want %v", event.IsTerminal(), c.IsTerminal)
			}

			// Re-serialize and deep-compare via parsed-JSON
			// normalization so int/float64 + map ordering don't
			// false-positive a mismatch.
			out, err := json.Marshal(event)
			if err != nil {
				t.Fatalf("marshal event: %v", err)
			}
			var got interface{}
			if err := json.Unmarshal(out, &got); err != nil {
				t.Fatalf("re-unmarshal: %v", err)
			}
			var want interface{}
			if err := json.Unmarshal(c.Wire, &want); err != nil {
				t.Fatalf("unmarshal wire (normalize): %v", err)
			}
			if !reflect.DeepEqual(got, want) {
				t.Errorf("round-tripped JSON differs from wire\n got:  %v\n want: %v", got, want)
			}
		})
	}
}

File: net/crates/net/bindings/node/test/tool_event_golden_vectors.test.ts
// Cross-language ToolEvent envelope round-trip fixture test (plan T-2).
//
// Loads `crates/net/tests/cross_lang_tool_formats/tool_event_vectors.json`
// and asserts that for each case the Node TS ToolEvent
// representation round-trips through JSON byte-equal to the wire
// shape. Matches the Rust verifier at
// `sdk/tests/tool_event_golden_vectors.rs`.

import { readFileSync } from 'node:fs'
import { join } from 'node:path'

import { describe, expect, it } from 'vitest'

import { isTerminalEvent, ToolEvent } from '../tool'

const FIXTURE_PATH = join(
  __dirname,
  '..',
  '..',
  '..',
  'tests',
  'cross_lang_tool_formats',
  'tool_event_vectors.json',
)

interface FixtureCase {
  name: string
  wire: Record<string, unknown>
  is_terminal: boolean
}

interface Fixture {
  cases: FixtureCase[]
}

const fixture: Fixture = JSON.parse(readFileSync(FIXTURE_PATH, 'utf8'))

describe('ToolEvent round-trip matches golden vectors', () => {
  for (const c of fixture.cases) {
    it(c.name, () => {
      // The Node TS ToolEvent is a tagged-union TypeScript type.
      // Its JSON representation IS the wire shape — no separate
      // serialize/deserialize step beyond the cast. Round-trip is
      // therefore JSON.parse(JSON.stringify(event)).
      const event = c.wire as unknown as ToolEvent
      // is_terminal contract
      expect(isTerminalEvent(event)).toBe(c.is_terminal)
      // Round-trip through JSON to assert no extra fields, no
      // dropped fields, no key reordering issues.
      const roundTripped = JSON.parse(JSON.stringify(event))
      expect(roundTripped).toEqual(c.wire)
    })
  }
})
